Split brain detection

sergepetrenko opened the pull request:
Split-brain machinery: filters that monitor PROMOTE, DEMOTE, CONFIRM, ROLLBACK coming from remote nodes, and raise an error, when the requests are conflicting with the node's local state.
Note, any asynchronous transaction coming from an obsolete journal term is treated as a sign of split-brain.

Closes #5295
Closes #6133

cyrillos commented:
Hi @sergepetrenko, great work! I think we should give it a spin asap, pretty critical feature, even if there are some small nits left we could fix them on top. ACK.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+	 * Wait until all the entries receive an lsn. The lsn will be
+	 * used to determine whether filtered request is safe to apply.
+	 */
+	if (txn_limbo_wait_persisted(limbo) < 0)
```
That is a bit worrying. We do that in the most common case when we are the limbo owner and are writing normal transactions with correct CONFIRM. I suppose when you are the limbo owner and you write CONFIRM yourself, then you don't need to wait for last rows to finish WAL write, don't you? Could it be kept like that in some simple way?

Gerold103 replied:
Oh, right, CONFIRM is written bypassing the txn_limbo_req flow (for now). Then should be fine atm.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+	 * correctly.
+	 */
+	assert(!limbo->is_in_rollback);
+	limbo->is_in_rollback = true;
```
Did I see correctly, that even if everything is fine and dandy, and we are writing a regular CONFIRM on the limbo owner itself, then new transactions will be rolled back on attempt to commit? Doesn't it look a bit hard to use if you can't safely write more than one synchro txn at a time or you will get rollbacks? Or did I misunderstand something?

sergepetrenko replied:
I re-checked, and it's fine.
Local CONFIRM and ROLLBACK bypass the filter completely. Just like they do with `txn_limbo_req_prepare`.

Gerold103 replied:
Good, goooood.

Gerold103 commented on test/replication-luatest/gh_5295_split_brain_test.lua:
```diff
@@ -0,0 +1,247 @@
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+
+local g = t.group('gh-6036')
```
6036 -> 5295.

sergepetrenko replied:
Oops. Thanks!

Gerold103 commented on test/replication-luatest/gh_5295_split_brain_test.lua:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    cg.s = {}
+    for i = 1,num_instances do
```
Lets scatter some whitespaces after commas in this file.

sergepetrenko replied:
As you wish. I thought `for j = 1,i` was an idiom in lua.

Gerold103 commented on test/replication-luatest/gh_5295_split_brain_test.lua:
```diff
@@ -0,0 +1,247 @@
+local t = require('luatest')
```
I've got this when run the test 30 times:
```
python test-run.py $(yes 5295 | head -n 30)
```
```
[012] replication-luatest/gh_5295_split_brain_test.l>                 [ fail ]
[012] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/gh_5295_split_brain.reject:
[012] TAP version 13
[012] 1..8
[012] # Started on Wed Jun 15 00:10:05 2022
[012] # Starting group: gh-6036
[012] not ok 1	gh-6036.test_async_old_term
[012] #   Peer closed
[012] #   stack traceback:
[012] #   	builtin/box/net_box.lua:747: in function '_request'
[012] #   	builtin/box/net_box.lua:807: in function 'eval'
[012] #   	...ol/test/replication-luatest/gh_5295_split_brain_test.lua:95: in function 'partition_nth_instance'
[012] #   	...ol/test/replication-luatest/gh_5295_split_brain_test.lua:136: in function 'gh-6036.test_async_old_term'
[012] #   	...
[012] #   	[C]: in function 'xpcall'
[012] s1 | Assertion failed: (replicaset.applier.synced > 0), function replica_on_applier_disconnect, file /Users/gerold/Work/Repositories/tarantool/src/box/replication.cc, line 498.
[012] not ok 2	gh-6036.test_confirm_old_term
```
Is it related to the patch? I didn't see you changing `synced` counter or anything related in this PR.

sergepetrenko replied:
Yep, I've seen it too. I don't know how it could be connected to the patch.
This might be some uncovered issue with starting a too big cluster.

On my side the issue disappeared when I removed the `replication_connect_quorum = 1` line from the node config.
Looks like I've simply decreased the chances to reproduce it.

I think this will disappear once I rework the test with only having 2 nodes in cluster at all times.

Gerold103 replied:
Lets pretend it never happened so we don't need to file a ticket.

sergepetrenko replied:
Yep. Looks like the issue reproduces when bootstrapping a rather big cluster. Say, 9 nodes. Like it was with previous test version.

Gerold103 commented on test/replication-luatest/gh_5295_split_brain_test.lua:
```diff
@@ -0,0 +1,247 @@
+end)
+
+g.before_each(function(cg)
+    -- Check that the servers start synced and with s[1] being leader.
+    -- It's a prerequisite for each test.
+    cg.s[1]:exec(function()
+        require('luatest').assert_equals(box.info.synchro.queue.owner,
+                                         box.info.id, 'First node is leader')
+    end)
+    t.helpers.retrying({}, function()
+        cg.s[num_instances]:assert_follows_upstream(cg.main_id)
+    end)
```
But you check it only for the last server: `cg.s[num_instances]`, don't you?

sergepetrenko replied:
Yes. I only care about the next testing server (`cg.s[num_instances]`) and its connection to the leader.
But the testing server changes before each test: there's `num_instances = num_instances - 1` in `g.after_each`.

Gerold103 commented on test/replication-luatest/gh_5295_split_brain_test.lua:
```diff
@@ -0,0 +1,247 @@
+-- server per test case. Once partitioned, create a split-brain situation
+-- between the partitioned node and all the main's followers. Then check that
+-- the partitioned node can't reconnect.
+local num_instances = 1
```
What if we eventually have more than 30 test case (+1 main instance, +1 is reserved for 0 vclock)? Maybe it would be more reliable and a bit easier globally to start one new instance in each case, drop it in the end, and unregister it from `_cluster`?

sergepetrenko replied:
Hmm. I don't know why I haven't come up with this myself.
I was too fascinated with my approach for some reason :)
Will do.

sergepetrenko replied:
I've reworked the test:
```diff
diff --git a/test/replication-luatest/gh_5295_split_brain_test.lua b/test/replication-luatest/gh_5295_split_brain_test.lua
index db5499016..554346489 100644
--- a/test/replication-luatest/gh_5295_split_brain_test.lua
+++ b/test/replication-luatest/gh_5295_split_brain_test.lua
@@ -4,29 +4,22 @@ local server = require('test.luatest_helpers.server')
 
 local g = t.group('gh-5295')
 
-local function repl_list(i)
-    local repl = {}
-    for j = 1, i do
-        table.insert(repl, server.build_instance_uri('s'..j))
-    end
-    return repl
-end
+-- To distinguish replicas by name
+local test_id = 0
 
 -- gh-5295: the server should stop replication from an upstream which sends data
 -- conflicting in terms of promote / confirmed_lsn.
 --
 -- The test requires one instance in replicaset per each split-brain case, plus
--- one "main" instance. The idea is to partition one instance from the main
--- server per test case. Once partitioned, create a split-brain situation
--- between the partitioned node and all the main's followers. Then check that
+-- one "main" instance. The idea of each test case is first to join the instance
+-- and then partition it from the main server. Once partitioned, create a
+-- split-brain situation between the partitioned node and main. Then check that
 -- the partitioned node can't reconnect.
-local num_instances = 1
 
 g.before_all(function(cg)
     cg.cluster = cluster:new({})
 
     cg.box_cfg = {
-        replication = repl_list(num_instances),
         replication_timeout         = 0.1,
         replication_synchro_quorum  = 1,
         replication_synchro_timeout = 0.01,
@@ -36,44 +29,63 @@ g.before_all(function(cg)
         log_level                   = 6,
     }
 
-    cg.s = {}
-    for i = 1, num_instances do
-        cg.s[i] = cg.cluster:build_and_add_server({
-            alias = 's'..i,
-            box_cfg = cg.box_cfg,
-        })
-    end
+    cg.main = cg.cluster:build_and_add_server{
+        alias = 'main',
+        box_cfg = cg.box_cfg,
+    }
     cg.cluster:start()
 
-    cg.main_id = cg.s[1]:exec(function()
+    cg.main:exec(function()
         box.ctl.promote()
         box.ctl.wait_rw()
         local s = box.schema.space.create('sync', {is_sync = true})
         s:create_index('pk')
         s = box.schema.space.create('async')
         s:create_index('pk')
-        return box.info.id
+        -- Check the test correctness.
+        assert(box.info.id == 1)
     end)
 end)
 
+local function update_replication(...)
+    return box.cfg{replication = {...}}
+end
+
 g.before_each(function(cg)
-    -- Check that the servers start synced and with s[1] being leader.
+    -- Check that the servers start synced and with main being leader.
     -- It's a prerequisite for each test.
-    cg.s[1]:exec(function()
+    cg.main:exec(function()
         require('luatest').assert_equals(box.info.synchro.queue.owner,
-                                         box.info.id, 'First node is leader')
+                                         box.info.id, 'main node is leader')
     end)
+
+    test_id = test_id + 1
+    cg.box_cfg.replication = {
+            server.build_instance_uri('main'),
+            server.build_instance_uri('split_replica'..test_id),
+    }
+    cg.split_replica = cg.cluster:build_and_add_server{
+        alias = 'split_replica'..test_id,
+        box_cfg = cg.box_cfg,
+    }
+    cg.split_replica:start()
     t.helpers.retrying({}, function()
-        cg.s[num_instances]:assert_follows_upstream(cg.main_id)
+        cg.split_replica:assert_follows_upstream(1)
+    end)
+
+    cg.main:exec(update_replication, cg.box_cfg.replication)
+    t.helpers.retrying({}, function()
+        cg.main:assert_follows_upstream(2)
     end)
 end)
 
 -- Drop the partitioned server after each case of split-brain.
 g.after_each(function(cg)
-    cg.s[num_instances]:stop()
-    cg.s[num_instances]:cleanup()
-    cg.cluster.servers[num_instances] = nil
-    num_instances = num_instances - 1
+    cg.split_replica:stop()
+    cg.split_replica:cleanup()
+    -- Drop the replica's cluster entry, so that next one receives same id.
+    cg.main:exec(function() box.space._cluster:delete{2} end)
+    cg.cluster.servers[2] = nil
 end)
 
 g.after_all(function(cg)
@@ -81,31 +93,23 @@ g.after_all(function(cg)
     cg.cluster.servers = nil
 end)
 
-local function update_replication(...)
-    return box.cfg{replication = {...}}
-end
-
-local function partition_nth_instance(cg, n)
+local function partition_replica(cg)
     -- Each partitioning starts on synced servers.
-    cg.s[n]:wait_vclock_of(cg.s[1])
-    cg.s[1]:wait_vclock_of(cg.s[n])
-    cg.s[n]:exec(update_replication, {})
-    cg.box_cfg.replication = repl_list(n - 1)
-    for i = 1, n - 1 do
-        cg.s[i]:exec(update_replication, cg.box_cfg.replication)
-    end
-    return cg.s[n]
+    cg.split_replica:wait_vclock_of(cg.main)
+    cg.main:wait_vclock_of(cg.split_replica)
+    cg.split_replica:exec(update_replication, {})
+    cg.main:exec(update_replication, {})
 end
 
-local function reconnect_and_check_split_brain(srv, main_id)
-    srv:exec(update_replication, {server.build_instance_uri('s1')})
-    t.helpers.retrying({}, srv.exec, srv, function(main_id)
-        local upstream = box.info.replication[main_id].upstream
+local function reconnect_and_check_split_brain(srv)
+    srv:exec(update_replication, {server.build_instance_uri('main')})
+    t.helpers.retrying({}, srv.exec, srv, function()
+        local upstream = box.info.replication[1].upstream
         local t = require('luatest')
         t.assert_equals(upstream.status, 'stopped', 'replication is stopped')
         t.assert_str_contains(upstream.message, 'Split-Brain discovered: ',
                               false, 'split-brain is discovered')
-    end, {main_id})
+    end)
 end
 
 local function write_promote()
@@ -130,81 +134,74 @@ local function write_demote()
     t.assert_equals(box.info.synchro.queue.owner, 0, "Demote succeeded")
 end
 
-num_instances = num_instances + 1
 -- Any async transaction performed in an obsolete term means a split-brain.
 g.test_async_old_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(function(n) box.space.async:replace{n} end, {num_instances})
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(function() box.space.async:replace{1} end)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Any unseen sync transaction confirmation from an obsolete term means a
 -- split-brain.
 g.test_confirm_old_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(function(n) box.space.sync:replace{n} end, {num_instances})
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(function() box.space.sync:replace{1} end)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Any unseen sync transaction rollback from an obsolete term means a
 -- split-brain.
 g.test_rollback_old_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(function(n)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(function()
         box.cfg{replication_synchro_quorum = 31}
-        pcall(box.space.sync.replace, box.space.sync, {n})
+        pcall(box.space.sync.replace, box.space.sync, {1})
         box.cfg{replication_synchro_quorum = 1}
-    end, {num_instances})
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    end)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Conflicting demote for the same term is a split-brain.
 g.test_demote_same_term = function(cg)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(write_demote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
-    cg.s[1]:exec(write_promote)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(write_demote)
+    reconnect_and_check_split_brain(cg.split_replica)
+    cg.main:exec(write_promote)
 end
 
-num_instances = num_instances + 1
 -- Conflicting promote for the same term is a split-brain.
 g.test_promote_same_term = function(cg)
-    cg.s[1]:exec(write_demote)
-    local test_server = partition_nth_instance(cg, num_instances)
-    test_server:exec(write_promote)
-    cg.s[1]:exec(write_promote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.main:exec(write_demote)
+    partition_replica(cg)
+    cg.split_replica:exec(write_promote)
+    cg.main:exec(write_promote)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Promote from a bigger term with lsn < confirmed_lsn is a split brain.
 g.test_promote_new_term_small_lsn = function(cg)
-    cg.s[num_instances]:exec(write_promote)
-    local test_server = partition_nth_instance(cg, num_instances)
-    assert(test_server == cg.s[num_instances]) -- Make sure the test is correct.
-    test_server:exec(function(n) box.space.sync:replace{n} end, {num_instances})
-    cg.s[1]:exec(write_promote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.split_replica:exec(write_promote)
+    partition_replica(cg)
+    cg.split_replica:exec(function() box.space.sync:replace{1} end)
+    cg.main:exec(write_promote)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-local function fill_queue_and_write(server, n)
-    local wal_write_count = server:exec(function(n)
+local function fill_queue_and_write(server)
+    local wal_write_count = server:exec(function()
         local fiber = require('fiber')
         box.cfg{
             replication_synchro_quorum = 31,
             replication_synchro_timeout = 1000,
         }
         local write_cnt = box.error.injection.get('ERRINJ_WAL_WRITE_COUNT')
-        fiber.new(box.space.sync.replace, box.space.sync, {n})
+        fiber.new(box.space.sync.replace, box.space.sync, {1})
         return write_cnt
-    end, {n})
+    end)
     t.helpers.retrying({}, server.exec, server, function(cnt)
         local t = require('luatest')
         local new_cnt = box.error.injection.get('ERRINJ_WAL_WRITE_COUNT')
@@ -221,27 +218,23 @@ local function perform_rollback(server)
     end)
 end
 
-num_instances = num_instances + 1
 -- Promote from a bigger term with lsn > confirmed_lsn is a split brain.
 g.test_promote_new_term_big_lsn = function(cg)
-    cg.s[num_instances]:exec(write_promote)
-    fill_queue_and_write(cg.s[num_instances], num_instances)
-    local test_server = partition_nth_instance(cg, num_instances)
-    assert(test_server == cg.s[num_instances])
-    perform_rollback(test_server)
-    cg.s[1]:exec(write_promote)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.split_replica:exec(write_promote)
+    fill_queue_and_write(cg.split_replica)
+    partition_replica(cg)
+    perform_rollback(cg.split_replica)
+    cg.main:exec(write_promote)
+    reconnect_and_check_split_brain(cg.split_replica)
 end
 
-num_instances = num_instances + 1
 -- Promote from a bigger term with conflicting queue contents is a split brain.
 g.test_promote_new_term_conflicting_queue = function(cg)
-    cg.s[num_instances]:exec(write_promote)
-    fill_queue_and_write(cg.s[num_instances], num_instances)
-    local test_server = partition_nth_instance(cg, num_instances)
-    assert(test_server == cg.s[num_instances])
-    perform_rollback(test_server)
-    cg.s[1]:exec(write_promote)
-    fill_queue_and_write(test_server, num_instances)
-    reconnect_and_check_split_brain(test_server, cg.main_id)
+    cg.split_replica:exec(write_promote)
+    fill_queue_and_write(cg.split_replica)
+    partition_replica(cg)
+    perform_rollback(cg.split_replica)
+    cg.main:exec(write_promote)
+    fill_queue_and_write(cg.split_replica)
+    reconnect_and_check_split_brain(cg.split_replica)
 end

```

Gerold103 commented on test/replication-luatest/gh_5295_split_brain_test.lua:
```diff
@@ -0,0 +1,247 @@
+num_instances = num_instances + 1
+-- Any async transaction performed in an obsolete term means a split-brain.
+g.test_async_old_term = function(cg)
+    local test_server = partition_nth_instance(cg, num_instances)
+    test_server:exec(write_promote)
+    cg.s[1]:exec(function(n) box.space.async:replace{n} end, {num_instances})
+    reconnect_and_check_split_brain(test_server, cg.main_id)
```
I don't think I understand. In the tests you separate the last instance s[n]. And promote it. So it gets term > the term of s[1]. Why does it even fail? Shouldn't s[1] just receive latest updates from s[n] and agree that the latter is a leader? And if it should fail somewhy, then why after reconnect s[1] stays fine? Shouldn't it also fail? Or it does fail its downstream to s[b] but it is not checked in the tests?

sergepetrenko replied:
I only connect s[n] to s[1]. Not vice versa. So s[1] remains intact, not knowing that anything bad has happened. All the split brain conditions are checked on s[n].
In this particular case (with async transaction), s[1] would probably follow s[n] just fine, indeed.

Do we need to do something about it? Can we do anything about it? All the split brain detection relies on limbo terms and confirmed_lsn. And neither of these are violated if you connect s[1] to s[n].

Now you might ask, why an async transaction in a old term means a split-brain at all. We came up with this while discussing fencing: https://github.com/tarantool/tarantool/issues/7110#issuecomment-1145868767

The idea is that once the node is not a leader, it shouldn't be able to spoil everyone else's life anyhow.
It wasn't true until we disallow async transactions in an old term.

Gerold103 replied:
Thanks for the explanation, I see now.

> Do we need to do something about it? Can we do anything about it? All the split brain detection relies on limbo terms and confirmed_lsn. And neither of these are violated if you connect s[1] to s[n].

If you want. I am fine with how it is now too.

Gerold103 reviewed (changes requested):
Thanks for finalizing this important patchset! Good job, really!

sergepetrenko commented on changelogs/unreleased/gh-5295-split-brain-detection.md:
```diff
@@ -0,0 +1,6 @@
+## feature/raft
+
```
Checkpatch doesn't see the doc request for some reason. But I couldn't understand what's wrong exactly.
I remember there was an issue with checkpatch, when it stopped checking the commit message right on `---`.
Probably because `---` is a separator between the commit message and the diff itself in `git format-patch` output.

I don't have `---` in the message, so there's some other issue. But I didn't bother looking for the exact line which causes checkpatch to stop

sergepetrenko commented on changelogs/unreleased/gh-5295-split-brain-detection.md:
```diff
@@ -0,0 +1,6 @@
+## feature/raft
```
Okay, fixed.

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -586,7 +586,8 @@ txn_limbo_ack(struct txn_limbo *limbo, uint32_t replica_id, int64_t lsn)
 		return;
-	if (limbo->frozen)
+	/** Do not ack anything while being read-only. */
+	if (limbo->frozen || box_is_ro())
```
> That looks dangerous. What if I set box.cfg.read_only to true after a transaction is started and then collect ACKs? No new leader and I could write a CONFIRM just fine, but here I won't. Also, when if I set it back to false afterwards, the transaction would be stuck and rolled back on timeout?

Good point. As discussed in the chat, I've reworked it so that limbo now has 2 reasons for being frozen:
1. fencing
2. recovering a limbo owner and waiting for at least one PROMOTE to happen

Bullet 2 covers the case I was trying to fix with `box_is_ro()`.

> Besides, the issue can happen even after recovery, if all local data is restored, but replication is still catching up with the other nodes. Isn't it true?

Agree.

> The behaviour of collecting ACKs after restart looks legitimate, I think its absence was even a bug at some point.

We agreed it was a bug in a private chat. The reasoning is in commit message for commit "txn_limbo: do not confirm/rollback anything after restart"

> Or do you mean CONFIRMs are being written right during recovery even if there are CONFIRMs written later in WAL? Then we could add a new function txn_limbo_recover(request), like raft_process_recovery(request) which solves a similar problem.

No. You understood me correctly from the beginning. I meant that the node recovers not-applied txns, and then starts collecting ACKs even before checking if anyone else has become the new leader

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -457,6 +457,13 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
+	 * comparing existing confirm_lsn with the one arriving from a remote
+	 * instance.
+	 */
+	if (limbo->confirmed_lsn < lsn)
+		limbo->confirmed_lsn = lsn;
```
It can't on a leader, but it can on a replica.
Since both replica and leader issue `txn_limbo_read_confirm`, it was easier for me to fix the replica this way.
In other words, this branch only works on everyone but the synchro queue owner.

Having explained this to you, I decided to refactor it this way:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index b9c0cf85c..dadabfc16 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -469,8 +469,10 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
         * comparing existing confirm_lsn with the one arriving from a remote
         * instance.
         */
-       if (limbo->confirmed_lsn < lsn)
+       if (limbo->owner_id != instance_id) {
+               assert(limbo->confirmed_lsn <= lsn);
                limbo->confirmed_lsn = lsn;
+       }
 }
 
 /**
```

With this change some replication tests on commit "txn_limbo: track CONFIRM lsn on replicas" fail with assertion due to these lines:
https://github.com/tarantool/tarantool/blob/b3f462bf442a98dd2ae61b6cd7f863d2e2e8c1e8/src/box/txn_limbo.c#L867-L883

They are obsolete in presence of split-brain detection, and everything works fine and passes once the commit "txn_limbo: filter incoming synchro requests" is applied

sergepetrenko replied:
Uh oh, this doesn't work, unfortunately. You also have to track confirmed_lsn while recovering an instance.

I tried the following diff, but it still fails the assertion occasionally.

```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index dadabfc16..f3c9bb1d8 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -465,11 +465,11 @@ txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn)
                txn_complete_success(e->txn);
        }
        /*
-        * Track CONFIRM lsn on replica in order to detect split-brain by
-        * comparing existing confirm_lsn with the one arriving from a remote
-        * instance.
+        * Track CONFIRM lsn on replica and on recovering queue owner in order
+        * to detect split-brain by comparing existing confirm_lsn with the
+        * one arriving from a remote instance.
         */
-       if (limbo->owner_id != instance_id) {
+       if (limbo->owner_id != instance_id || !box_is_configured()) {
                assert(limbo->confirmed_lsn <= lsn);
                limbo->confirmed_lsn = lsn;
        }
```

I don't have any ideas on how to fix this now, so returned to the original version.

Gerold103 replied:
Hm, but how can it be less than the last received one on the replicas?

sergepetrenko replied:
Oh. Now I get your question.
When running this code (`if confirmed_lsn < lsn`) on replica, it's only possible for `confirmed_lsn` to be equal to `lsn` while reading a PROMOTE or DEMOTE. `txn_limbo_req_prepare` sets limbo->confirmed_lsn to request lsn, and then we enter `txn_limbo_req_commit`, `txn_limbo_read_promote`. The latter in turn calls `txn_limbo_read_confirm`, where `confirmed_lsn == lsn`.

 It's never possible that `confirmed_lsn > lsn` on a replica. Unless we receive a broken packet.

Now, as I understand, there are some cases when `confirmed_lsn > lsn` when recovering a limbo owner.
I tried to avoid the assertions by adding `if owner_id != 0 && owner_id != instance_id` but that didn't help either.

This is where I stopped investigating and returned to the original simple version with
`if confirmed_lsn < lsn then confirmed_lsn = lsn`.

So, short answer is: lsn is always >= confirmed_lsn on a replica, but it seems the opposite might happen on master during recovery.

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -216,6 +218,9 @@ txn_limbo_assign_lsn(struct txn_limbo *limbo, struct txn_limbo_entry *entry,
 	else
 		txn_limbo_assign_remote_lsn(limbo, entry, lsn);
+	if (entry == txn_limbo_last_entry(limbo)) {
+		fiber_cond_broadcast(&limbo->persist_cond);
```
Seems like it would be possible. Do you want me to try it?

I like the current approach, TBH. Why do you think `txn_on_wal_write` would be better?

Gerold103 replied:
I simply don't like introducing new entities when job can be done with the existing ones. You can try if you want. I can live with this version too.

sergepetrenko replied:
Ok. Turned out pretty easy:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 701229598..c5621dcd9 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -46,7 +46,6 @@ txn_limbo_create(struct txn_limbo *limbo)
 	limbo->len = 0;
 	limbo->owner_id = REPLICA_ID_NIL;
 	fiber_cond_create(&limbo->wait_cond);
-	fiber_cond_create(&limbo->persist_cond);
 	vclock_create(&limbo->vclock);
 	vclock_create(&limbo->promote_term_map);
 	limbo->promote_greatest_term = 0;
@@ -226,9 +225,6 @@ txn_limbo_assign_lsn(struct txn_limbo *limbo, struct txn_limbo_entry *entry,
 		txn_limbo_assign_local_lsn(limbo, entry, lsn);
 	else
 		txn_limbo_assign_remote_lsn(limbo, entry, lsn);
-	if (entry == txn_limbo_last_entry(limbo)) {
-		fiber_cond_broadcast(&limbo->persist_cond);
-	}
 }
 
 static void
@@ -789,6 +785,15 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
 	return 0;
 }
 
+static int
+txn_write_cb(struct trigger *trigger, void *event)
+{
+	(void)event;
+	struct fiber *fiber = (struct fiber *)trigger->data;
+	fiber_wakeup(fiber);
+	return 0;
+}
+
 /**
  * Wait until all the limbo entries receive an lsn.
  */
@@ -797,13 +802,18 @@ txn_limbo_wait_persisted(struct txn_limbo *limbo)
 {
 	if (txn_limbo_is_empty(limbo))
 		return 0;
-
-	while (txn_limbo_last_entry(limbo)->lsn <= 0) {
-		fiber_cond_wait(&limbo->persist_cond);
+	struct txn_limbo_entry *e = txn_limbo_last_entry(limbo);
+	while (e != NULL && e->lsn <= 0) {
+		struct trigger on_wal_write;
+		trigger_create(&on_wal_write, txn_write_cb, fiber(), NULL);
+		txn_on_wal_write(e->txn, &on_wal_write);
+		fiber_yield();
+		trigger_clear(&on_wal_write);
 		if (fiber_is_cancelled()) {
 			diag_set(FiberIsCancelled);
 			return -1;
 		}
+		e = txn_limbo_last_entry(limbo);
 	}
 	return 0;
 }
diff --git a/src/box/txn_limbo.h b/src/box/txn_limbo.h
index 05104d2ab..9ab67a5e7 100644
--- a/src/box/txn_limbo.h
+++ b/src/box/txn_limbo.h
@@ -116,11 +116,6 @@ struct txn_limbo {
 	 * the parameters are updated.
 	 */
 	struct fiber_cond wait_cond;
-	/**
-	 * A condition variable to wait until all the limbo entries are
-	 * persisted (but not necessarily completed).
-	 */
-	struct fiber_cond persist_cond;
 	/**
 	 * All components of the vclock are versions of the limbo
 	 * owner's LSN, how it is visible on other nodes. For

```

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+	return 0;
+}
+
+/**
+ * Fill the reject reason with request data.
+ * The function is not reenterable, use with care.
+ */
+static char *
+reject_str(const struct synchro_request *req)
+{
+	const char *type_name = iproto_type_name(req->type);
+	static char prefix[128];
+
+	snprintf(prefix, sizeof(prefix), "RAFT: rejecting %s (%d) "
+		 "request from origin_id %u replica_id %u term %llu",
+		 type_name ? type_name : "UNKNOWN", req->type,
+		 req->origin_id, req->replica_id,
+		 (long long)req->term);
+
+	return prefix;
+}
```
Sure. I missed that. Thanks!
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index fce9d74ba..9c08cb652 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -35,6 +35,7 @@
 #include "journal.h"
 #include "box.h"
 #include "raft.h"
+#include "tt_static.h"
 
 struct txn_limbo txn_limbo;
 
@@ -813,19 +814,16 @@ txn_limbo_wait_persisted(struct txn_limbo *limbo)
  * Fill the reject reason with request data.
  * The function is not reenterable, use with care.
  */
-static char *
+static const char *
 reject_str(const struct synchro_request *req)
 {
        const char *type_name = iproto_type_name(req->type);
-       static char prefix[128];
 
-       snprintf(prefix, sizeof(prefix), "RAFT: rejecting %s (%d) "
-                "request from origin_id %u replica_id %u term %llu",
-                type_name ? type_name : "UNKNOWN", req->type,
-                req->origin_id, req->replica_id,
-                (long long)req->term);
-
-       return prefix;
+       return tt_sprintf("RAFT: rejecting %s (%d) request from origin_id %u "
+                         "replica_id %u term %llu",
+                         type_name ? type_name : "UNKNOWN", req->type,
+                         req->origin_id, req->replica_id,
+                         (long long)req->term);
 }
 
 /**
```

cyrillos replied:
I've been considering `tt_sprintf` in first place then dropped this idea because I wanted the reject string to be independent of any other helpers. Though I'm fine with it.

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+ * over a valid lsn range.
+ */
+static int
+filter_queue_boundaries(struct txn_limbo *limbo,
```
Sure:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 9c08cb652..73cddd596 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -871,8 +871,8 @@ txn_limbo_filter_generic(struct txn_limbo *limbo,
  * over a valid lsn range.
  */
 static int
-filter_queue_boundaries(struct txn_limbo *limbo,
-                       const struct synchro_request *req)
+txn_limbo_filter_queue_boundaries(struct txn_limbo *limbo,
+                                 const struct synchro_request *req)
 {
        int64_t lsn = req->lsn;
        /*
@@ -942,8 +942,8 @@ filter_queue_boundaries(struct txn_limbo *limbo,
  * Filter CONFIRM and ROLLBACK packets.
  */
 static int
-filter_confirm_rollback(struct txn_limbo *limbo,
-                       const struct synchro_request *req)
+txn_limbo_filter_confirm_rollback(struct txn_limbo *limbo,
+                                 const struct synchro_request *req)
 {
        assert(latch_is_locked(&limbo->promote_latch));
        assert(limbo->do_validate);
@@ -958,13 +958,13 @@ filter_confirm_rollback(struct txn_limbo *limbo,
                return -1;
        }
 
-       return filter_queue_boundaries(limbo, req);
+       return txn_limbo_filter_queue_boundaries(limbo, req);
 }
 
 /** A filter PROMOTE and DEMOTE packets. */
 static int
-filter_promote_demote(struct txn_limbo *limbo,
-                     const struct synchro_request *req)
+txn_limbo_filter_promote_demote(struct txn_limbo *limbo,
+                               const struct synchro_request *req)
 {
        assert(latch_is_locked(&limbo->promote_latch));
        assert(limbo->do_validate);
@@ -994,7 +994,7 @@ filter_promote_demote(struct txn_limbo *limbo,
                return -1;
        }
 
-       return filter_queue_boundaries(limbo, req);
+       return txn_limbo_filter_queue_boundaries(limbo, req);
 }
 
 /** A fine-grained filter checking specific request type constraints. */
@@ -1013,10 +1013,10 @@ txn_limbo_filter_request(struct txn_limbo *limbo,
        switch (req->type) {
        case IPROTO_RAFT_CONFIRM:
        case IPROTO_RAFT_ROLLBACK:
-               return filter_confirm_rollback(limbo, req);
+               return txn_limbo_filter_confirm_rollback(limbo, req);
        case IPROTO_RAFT_PROMOTE:
        case IPROTO_RAFT_DEMOTE:
-               return filter_promote_demote(limbo, req);
+               return txn_limbo_filter_promote_demote(limbo, req);
        default:
                unreachable();
        }

```

cyrillos replied:
Guys, seriously, this is over the top already. `txn_limbo_filter_promote_demote` is 31 chars for **static function** name?! It is not exported anywhere due to file scope visibility, please keep things simple. This has nothing to do with _methods_.

sergepetrenko replied:
My thinking was similar: it's `txn_limbo_filter_smth` for public methods and simply `filter_smth` for static ones.
Anyway, it's a matter of seconds to revert.

Gerold103 replied:
> txn_limbo_filter_promote_demote is 31 chars for static function name?!

OH, NOOOOO, 31 CHARS!!!😱 IN A STATIC FUNCTION!!!😱😱

> It is not exported anywhere due to file scope visibility, please keep things simple.

Let me clarify. If you would need to expose that function into the header at some point in the future, you will rename it? And if a function becomes static, you will rename it back? I wouldn't even mind if all the filter functions would be static, then all of them could have filter_ prefix. But here you added inconsistency. Some filtering functions are methods of the limbo and have the prefix, the others do not. That is not right.

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+ */
+static int
+filter_confirm_rollback(struct txn_limbo *limbo,
+			const struct synchro_request *req)
```
No problem:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 73cddd596..bbdf1a983 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -947,7 +947,8 @@ txn_limbo_filter_confirm_rollback(struct txn_limbo *limbo,
 {
        assert(latch_is_locked(&limbo->promote_latch));
        assert(limbo->do_validate);
-
+       assert(req->type == IPROTO_RAFT_CONFIRM ||
+              req->type == IPROTO_RAFT_ROLLBACK);
        /*
         * Zero LSN is allowed for PROMOTE and DEMOTE requests only.
         */
@@ -968,7 +969,7 @@ txn_limbo_filter_promote_demote(struct txn_limbo *limbo,
 {
        assert(latch_is_locked(&limbo->promote_latch));
        assert(limbo->do_validate);
-
+       assert(iproto_type_is_promote_request(req->type));
        /*
         * PROMOTE and DEMOTE packets must not have zero
         * term supplied, otherwise it is a broken packet.
```

sergepetrenko commented on src/box/txn_limbo.c:
```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+	 * Explicit split brain situation. Request comes in with an old LSN
+	 * which we've already processed.
+	 */
+	if (limbo->confirmed_lsn > lsn) {
```
Not all `if`s correspond to a reproducible scenario. Some of them (like the ones with lsn == 0, zero term, zero replica id, CONFIRM for lsn == confirmed_lsn) could only be reproduced with fuzzing.
Since these situations are not really a split brain, but some other issue, I tried to raise `ER_UNSUPPORTED` and not `ER_SPLIT_BRAIN` for them.

But speaking of all the `ER_SPLIT_BRAIN` errors, they are covered. At least they were until the todays push. Let's see how it goes now. I checked via coveralls.

sergepetrenko commented:
@Gerold103, please, take a look once again.
I've fixed everything but your suggestion with tests (to start only 2 instances, instead of number_of_test_cases + 1).
I'll follow up on that a bit later.
Everything else is ready for review.

kostja commented on changelogs/unreleased/gh-5295-split-brain-detection.md:
```diff
@@ -0,0 +1,9 @@
+
+* Fixed a possible split-brain when old synchro queue owner might finalize the
+  transactions in presence of a new synchro queue owner (gh-5295).
+
```
This is not split-brain. I wonder why you call it this way. Split brain is a specific case of network partitiong when neither side has a majority. It's impossible to detect. So you can't really have a split brain detection.

Gerold103 commented on src/box/txn_limbo.c:
```diff
@@ -52,13 +52,19 @@ txn_limbo_create(struct txn_limbo *limbo)
+static inline bool
+txn_limbo_is_frozen(const struct txn_limbo *limbo)
+{
+	return limbo->is_frozen_reasons != 0;
```
* Perhaps it is worth renaming `txn_limbo_is_frozen` to `txn_limbo_is_fenced` then.
* Sorry for the stupid nit, but since this is no longer a flag, it should not have `is_` prefix. Rather would be `fence_reason_mask` or at least `fence_reasons` (or frozen instead of fenced if you want).

Gerold103 commented on src/box/txn_limbo.h:
```diff
@@ -189,14 +189,26 @@ struct txn_limbo {
+			 * leader again or PROMOTE/DEMOTE arrives from some
+			 * remote instance.
+			 */
+			bool is_frozen_due_to_fencing : 1;
```
Core team will smack our asses for bit fields 🌚🌚🌚🌚🌚🌚 (fine for me).

sergepetrenko replied:
Let it be our little secret.

Gerold103 commented on src/box/raft.c:
```diff
@@ -201,6 +201,14 @@ box_raft_on_update_f(struct trigger *trigger, void *event)
+	 */
+	if (raft->volatile_term > txn_limbo.promote_greatest_term) {
+		txn_limbo_fence(&txn_limbo);
+	}
```
Nit: you don't need `{}` for one-line ifs here and in other places in other commits. Up to you anyway.

sergepetrenko replied:
Sure. Fixed.

Gerold103 reviewed (commented):
Good, very well! Splendid stuff, only a few things to iron out in the old comments and a few new ones.

sergepetrenko commented:
Thanks for the review, @Gerold103!
I've reworked the test as discussed and answered your comments.
Also I've fixed the #6133 issue in a separate commit on top, while we are at it.
Please, take a look.

Gerold103 reviewed (approved):
Thanks for the good patch! I didn't review the last version from top to bottom from scratch, but looks good to me.

kyukhin commented:
Backported to 2.10.
